core/extensions/impl/memory_extension.cpp

```diff
@@ -6,13 +6,34 @@
 #include <exception>
 
 #include "extensions/impl/memory_extension.hpp"
+#include "memory_extension.hpp"
 
 namespace kagome::extensions {
-  uint8_t *MemoryExtension::ext_malloc(uint32_t size) {
-    std::terminate();
+  MemoryExtension::MemoryExtension()
+      : logger_{common::createLogger("WASM Runtime [MemoryExtension]")} {}
+
+  MemoryExtension::MemoryExtension(std::shared_ptr<runtime::Memory> memory)
+      : memory_(std::move(memory)) {}
+
+  int32_t MemoryExtension::ext_malloc(uint32_t size) {
+    auto opt_ptr = memory_->allocate(size);
```

Comment:
why allocate returns optional? it can safely return -1 or 0 (nullptr), so this method will be one-liner

given type class Pointer, you can define special state, when Pointer is not allocated